examples/token/mint/src/main.rs

```diff
@@ -0,0 +1,103 @@
+//! Token mint
+//!
+//! It allows exchanging SMH for the token,
+//! which is then transfered to the chosen
+//! wallet address.
+//!
+//! Every mint account instance represents a
+//! unique token.
+#![no_main]
+
+use athena_interface::{Address, Encode, MethodSelector};
+use athena_vm::entrypoint;
+use athena_vm_declare::{callable, template};
+use athena_vm_sdk::{wallet::SpendArguments, Pubkey};
+use parity_scale_codec::Decode;
+
+/// Storage for the number of already distributed tokens
+const SUPPLY_KEY: [u32; 8] = [0; 8];
+const TOKEN_IDENTIFIER_KEY: [u32; 8] = [1; 8];
+
+#[derive(Debug, Decode, Encode)]
+struct Mint {
+  owner: Pubkey,
+  max_supply: u64,
+  /// token/smidge ratio (how many smidges per token)
+  token_price: u64,
+}
+
+athena_vm::entrypoint!();
+
+#[template]
+impl Mint {
+  #[callable]
+  fn spawn(mint: Mint) -> Address {
+    let address = athena_vm_sdk::spawn(&mint.encode());
+
+    let mut token_identifier_words = [0u32; 8];
+    for (i, c) in address.0.chunks_exact(4).enumerate() {
+      token_identifier_words[i] = u32::from_le_bytes(c.try_into().unwrap());
+    }
+    athena_vm::syscalls::write_storage(&TOKEN_IDENTIFIER_KEY, &token_identifier_words);
+    address
```

Comment:
Is the address only accessible during spawn? Wouldn't it be more efficient to read it from the host when needed instead of keeping it stored in the mutable state?

Reply:
Right. I updated the `Context` struct to keep the `callee` address as well and read it from there (instead of the mutable storage).